### ISSUE TITLE
Problem with using async and exec

### ISSUE THREAD

**Issue by santhosh-kumar** [OpenSIPS/opensips#829]
We are using async and exec to make a web-request as shown below:

        $var(fullQueryClientReadyRequest) = "an url with params";
        async(exec("curl --connect-timeout 30 --max-time 30 \"$var(fullQueryClientReadyRequest)\"", "", "$var(restGetResponse)"), RT_RESUME_AFTER_RESPONSE);

We see the following (printed several times):

> osips[16437]: NOTICE:core:io_wait_loop_epoll: connection closed by the remote peer!
> osips[16437]: ERROR:core:io_wait_loop_epoll: [UDP_worker] unexpected event 10 on 1/1, data=0x7f215063bb68

After this happens, the process becomes unresponsive.

**Note:** We do not use $var(restGetResponse) in the RT_RESUME_AFTER_RESPONSE. However, we have it so that the exec will be blocking.

**Comment by bogdan-iancu** [comment]
Hi @santhosh-kumar , this is very strange (the chain of errors you get) - it shows like the epoll reactor triggered only EPOLLHUP event, without any IN or OUT notification. Do you have any idea what happens at the network level with the corresponding TCP connection ? is there any connect attempt? or a reject from the web server ?

Thanks,
Bogdan

**Comment by santhosh-kumar** [comment]
Hi @bogdan-iancu,

Yes, there was a connection attempt in both the failure and success cases. So more information:

a) Curl makes a webrequest. Webservice will block until a timeout say 4s. Hence we used async for scalability.

**Success Case:**

```
osips[6987]: RequestRoute[RT_HANDLE_PSTN_REDIRECT]: Submitting web service get
osips[6987]: NOTICE:core:io_wait_loop_epoll: EPOLLIN(read) event: epollwait() set event EPOLLHUP - connection closed by the remote peer!
osips[6987]: ResumeRoute[RT_RESUME_AFTER_RESPONSE]: Async web service request returned

```
**Failure Case:**

```
osips[6611]: RequestRoute[RT_HANDLE_PSTN_REDIRECT]: Submitting web service get request
osips[6611]: NOTICE:core:io_wait_loop_epoll: connection closed by the remote peer!
osips[6611]: ERROR:core:io_wait_loop_epoll: [UDP_worker] unexpected event 10 on 1/1, data=0x7f8bce9f3b50
osips[6611]: NOTICE:core:io_wait_loop_epoll: connection closed by the remote peer!
osips[6611]: ERROR:core:io_wait_loop_epoll: [UDP_worker] unexpected event 10 on 1/1, data=0x7f8bce9f3b50
```

In both the cases, I verified using wireshark that the tcp connection was made with the server and server replied back with a response:

`{"result":"timeout"}`

**Comment by bogdan-iancu** [comment]
Thank you @santhosh-kumar . In the Failure Case, do you have pcap showing what happens at network level with the TCP connection ?

**Comment by santhosh-kumar** [comment]
Here is the dropbox link to the pcap: https://www.dropbox.com/s/jod9zdifa93d5yv/trace_2016-03-16_20_55-UTC.pcap?dl=0

It has webrequests made to port 80.

Thank you.

**Comment by bogdan-iancu** [comment]
Thank you @santhosh-kumar . There are 2 HTTP flows there, one IPv6 with 46732 port (/phoneNotification/...) and another one with IPV4 and port 49864  (/ws/index.php/...).
Any idea which is the faulty one ?

**Comment by santhosh-kumar** [comment]
2nd one is the faulty one i.e., port 49864 (/ws/index.php/...). I will be deleting the content of the dropbox since you downloaded it.

**Comment by santhosh-kumar** [comment]
**More information:**

I tried the following:
`
async(exec("sleep 4", "", "$var(restGetResponse)"), RT_RESUME_AFTER_RESPONSE);`

I still the error. Makes me wonder whether async+exec is the problem here.

**Comment by bogdan-iancu** [comment]
As per trace, both queries were successful (as HTTP), right ? If you can reproduce this, may provide a patch with some extra logs ?

**Comment by santhosh-kumar** [comment]
Yes, I can still reproduce it. 
**Note: **First query is done independently. The second query i.e., port 49864 (/ws/index.php/...) is causing this problem.

If you can provide me a patch with extra logging to the module, I can experiment and get back to you.

**Comment by bogdan-iancu** [comment]
Please apply this patch https://gist.github.com/bogdan-iancu/a20e800fbfc0477db138 . It will print as NOTICE the operations related to adding, removing or triggering inside the reactor.
Run a the failed scenario and send me the logs from IO Reactor (all the notice msgs).
Thanks, Bogdan

**Comment by santhosh-kumar** [comment]
Here is the log file with additional logging requested: https://www.dropbox.com/sh/93j7gdfwk2h5lnq/AACGTMrj6zKaXHjw_YBQBm84a?dl=0

I was able to reproduce both successful (good_async.txt) and failure (bad_async.txt) cases.

Please search for in the logs to identify the async+exec+curl web request:
`
Submitting web service get request - http://255.255.255.29:80/ws/index.php/query_client_ready_for_call`

Also, I found that in the failure case, the following log statement is missing:
```

Mar 22 19:48:27  osips[14660]: NOTICE:core:io_wait_loop_epoll: EPOLLIN(read) event: epollwait() set event EPOLLHUP - connection closed by the remote peer!
Mar 22 19:48:27  osips[14660]: NOTICE:tm:io_watch_del: [UDP_worker] io_watch_del op on index -1 37 (0x848de0, 37, -1, 0x10,0x1) fd_no=3 called
```

Thanks for the bug triage!

**Comment by bogdan-iancu** [comment]
@santhosh-kumar , there was an issue in the prev patch, please revert it and use this new one:
    https://gist.github.com/bogdan-iancu/fa596f4c99cda796e90a
Also, looking for logs :)

**Comment by santhosh-kumar** [comment]
Try this: https://www.dropbox.com/sh/93j7gdfwk2h5lnq/AACGTMrj6zKaXHjw_YBQBm84a?dl=0

I wasn't able to reproduce the successful case this time. It's failing all the time.

**Comment by bogdan-iancu** [comment]
@santhosh-kumar , many thanks for your help in troubleshooting this - I prepared a fix for this. Please remove the previous patches and apply (backport) this 2503bc4. Let me know if it fixes your problem.

**Comment by santhosh-kumar** [comment]
Thanks for the fix, I will try out and update you.

**Comment by santhosh-kumar** [comment]
@bogdan-iancu The fix seems to work fine for us. The error is gone now.

**Comment by bogdan-iancu** [comment]
Thank you @santhosh-kumar , I pushed the fix to 2.1 too !
